File: scripts/publish_guardrail_event.py
#!/usr/bin/env python3
import json
import os
import sys

from confluent_kafka import Producer

BOOTSTRAP = os.environ.get("KAFKA_BOOTSTRAP_SERVERS", "kafka:9092")
TOPIC = os.environ.get("KAFKA_TOPIC", "security.dependency.events")
RESULT_FILE = os.environ.get("GUARDRAIL_RESULT_FILE", "guardrail-result.json")
PIPELINE_ID = os.environ.get("CI_PIPELINE_ID", "unknown")
REPO = os.environ.get("CI_REPOSITORY", "unknown")
COMMIT_SHA = os.environ.get("CI_COMMIT_SHA", "unknown")


def delivery_report(err, msg):
    if err:
        print(f"Kafka delivery failed: {err}", file=sys.stderr)


def main() -> int:
    with open(RESULT_FILE, "r", encoding="utf-8") as f:
        result = json.load(f)

    payload = {
        "pipeline_id": PIPELINE_ID,
        "repository": REPO,
        "commit_sha": COMMIT_SHA,
        "status": result["status"],
        "summary": result.get("summary", {}),
        "blocked": result.get("blocked", []),
        "quarantined": result.get("quarantined", []),
    }

    producer = Producer({"bootstrap.servers": BOOTSTRAP})
    producer.produce(TOPIC, json.dumps(payload).encode("utf-8"), callback=delivery_report)
    producer.flush()

    print(f"Published guardrail event to {TOPIC}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/publish-guardrail-event.js
#!/usr/bin/env node
import fs from 'node:fs/promises';
import path from 'node:path';
import { Kafka } from 'kafkajs';

import { runWithObservability } from '../src/lib/observability.js';
import { buildGuardrailEventPayload } from '../src/lib/security-events.js';

const BOOTSTRAP = process.env.KAFKA_BOOTSTRAP_SERVERS || 'kafka:9092';
const TOPIC = process.env.KAFKA_TOPIC || 'security.dependency.events';
const RESULT_FILE = process.env.GUARDRAIL_RESULT_FILE || 'guardrail-result.json';
const PIPELINE_ID = process.env.CI_PIPELINE_ID || 'unknown';
const REPOSITORY = process.env.CI_REPOSITORY || 'unknown';
const COMMIT_SHA = process.env.CI_COMMIT_SHA || 'unknown';

const outcome = await runWithObservability({
  tool: 'publish_guardrail_event',
  execute: async () => {
    const resultPath = path.resolve(RESULT_FILE);
    const result = JSON.parse(await fs.readFile(resultPath, 'utf8'));
    const payload = buildGuardrailEventPayload({
      result,
      pipelineId: PIPELINE_ID,
      repository: REPOSITORY,
      commitSha: COMMIT_SHA,
    });

    const kafka = new Kafka({ brokers: BOOTSTRAP.split(',').map((item) => item.trim()).filter(Boolean) });
    const producer = kafka.producer();

    await producer.connect();
    await producer.send({ topic: TOPIC, messages: [{ value: JSON.stringify(payload) }] });
    await producer.disconnect();

    process.stdout.write(`Published guardrail event to ${TOPIC}\n`);

    return {
      exitCode: 0,
      metrics: {
        status: 'published',
        topic: TOPIC,
        event_status: payload.status || 'unknown',
      },
    };
  },
});

process.exit(outcome.exitCode);
